File: js/src/maplestory/api/dto/character/characterPetEquipmentItemDto.ts
import {
  CharacterPetEquipmentItemOptionDto
} from '@src/maplestory/api/dto/character/characterPetEquipmentItemOptionDto';
import { CharacterPetEquipmentItemDtoBody } from '@src/maplestory/api/response/character/characterPetEquipmentDtoBody';

/**
 * 캐릭터 펫 장비 아이템 정보
 */
class CharacterPetEquipmentItemDto {
  /**
   * 아이템 명
   */
  itemName: string | null;

  /**
   * 아이템 아이콘
   */
  itemIcon: string | null;

  /**
   * 아이템 설명
   */
  itemDescription: string | null;

  /**
   * 아이템 표기상 옵션
   */
  itemOption: CharacterPetEquipmentItemOptionDto[];

  /**
   * 업그레이드 횟수
   */
  scrollUpgrade: number;

  /**
   * 업그레이드 가능 횟수
   */
  scrollUpgradable: number;

  /**
   * 아이템 외형
   */
  itemShape: string | null;

  /**
   * 아이템 외형 아이콘
   */
  itemShapeIcon: string | null;

  constructor(obj: CharacterPetEquipmentItemDtoBody) {
    const {
      item_name,
      item_icon,
      item_description,
      item_option,
      scroll_upgrade,
      scroll_upgradable,
      item_shape,
      item_shape_icon
    } = obj;

    this.itemName = item_name;
    this.itemIcon = item_icon;
    this.itemDescription = item_description;
    this.itemOption = item_option.map(
      (option) => new CharacterPetEquipmentItemOptionDto(option),
    );
    this.scrollUpgrade = scroll_upgrade;
    this.scrollUpgradable = scroll_upgradable;
    this.itemShape = item_shape;
    this.itemShapeIcon = item_shape_icon;
  }
}

export { CharacterPetEquipmentItemDto };

File: java/src/main/java/dev/spiralmoon/maplestory/api/msea/dto/character/CharacterPetEquipmentItemDTO.java
package dev.spiralmoon.maplestory.api.msea.dto.character;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.ToString;

import java.util.List;

/**
 * Equipment information for pet
 */
@AllArgsConstructor
@NoArgsConstructor
@Data
@ToString
public class CharacterPetEquipmentItemDTO implements dev.spiralmoon.maplestory.api.common.dto.character.CharacterPetEquipmentItemDTO<
        CharacterPetEquipmentItemOptionDTO> {

    /**
     * Item name
     */
    @JsonProperty("item_name")
    private String itemName;

    /**
     * Item icon
     */
    @JsonProperty("item_icon")
    private String itemIcon;

    /**
     * Item description
     */
    @JsonProperty("item_description")
    private String itemDescription;

    /**
     * Displayed item options
     */
    @JsonProperty("item_option")
    @JsonSetter(nulls = Nulls.AS_EMPTY)
    private List<CharacterPetEquipmentItemOptionDTO> itemOption;

    /**
     * Number of upgrades
     */
    @JsonProperty("scroll_upgrade")
    private int scrollUpgrade;

    /**
     * Number of upgrades available
     */
    @JsonProperty("scroll_upgradable")
    private int scrollUpgradable;

    /**
     * Item appearance
     */
    @JsonProperty("item_shape")
    private String itemShape;

    /**
     * Item appearance icon
     */
    @JsonProperty("item_shape_icon")
    private String itemShapeIcon;
}
